test/core/xds/xds_client_ads_stream_wait_test.cc

```diff
@@ -0,0 +1,108 @@
+//
+// Copyright 2023 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include <memory>
+#include <string_view>
+
+#include "absl/status/status.h"
+#include "absl/types/optional.h"
+#include "gtest/gtest.h"
+
+#include <grpc/grpc.h>
+
+#include "src/core/lib/gprpp/ref_counted_ptr.h"
+#include "src/proto/grpc/testing/xds/v3/discovery.pb.h"
+#include "test/core/util/test_config.h"
+#include "test/core/xds/xds_client_test_lib.h"
+
+namespace grpc_core {
+namespace testing {
+namespace {
+
+using XdsClientNotifyWatchersDone = XdsClientTestBase;
+
+TEST_F(XdsClientNotifyWatchersDone, Basic) {
+  InitXdsClient();
+  // Start a watch for "foo1".
+  auto watcher = StartFooWatch("foo1");
```

Comment:
I suggest starting watchers for two different resources and having them both in the same response, so that we can verify that we don't start the next read until *both* watchers free their read delay handles.